### PR TITLE
rust: replace bash script with build script

### PR DESCRIPTION
**Pull request by nrc**
With this change, preparing the Rust code becomes the idiomatic `cargo build`, rather than requiring a script. We would like to migrate from rust-protobuf to Prost, and this change should make that easier.

The Cargo build script is pretty much a 1:1 re-implementation of the old script.

cc @ice1000

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -2,12 +2,20 @@
 name = "kvproto"
 version = "0.0.1"
 keywords = ["TiKV"]
+edition = "2018"
```
Comment (BusyJay) [inline Cargo.toml]: Is it OK that dependant project is not written in 2018 edition?

Comment (nrc) [reply Cargo.toml]: Yes, it's fine. The editions are designed so that a 2018 crate can depend on a 2015 crate and a 2015 crate can depend on a 2018 crate.

**Comment by siddontang** [comment]
LGTM

PTAL @BusyJay

File: build.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            let file_name = e.expect("Couldn't list file").file_name();
+            file_name
+                .to_string_lossy()
+                .split(".rs")
+                .next()
+                .map(|n| n.to_owned())
+        })
+        .collect();
+    
+    replace_read_unknown_fields(&mod_names);
```
Comment (BusyJay) [inline build.rs]: How about sort it first?

Comment (nrc) [reply build.rs]: Do you mean sort `mod_names` so that the `mod` declarations in lib.rs are in alphabetical order? Or for some other purpose?

Comment (BusyJay) [reply build.rs]: Yes, exactly.

Comment (ice1000) [reply build.rs]: @BusyJay I think this PR is kind of obsolete due to  #349.

**Comment by BusyJay** [comment]
Another thought: it may be bad to modify the source directory. It's more natural to generate sources into the target build directory and link it into source directory. For example, for now whenever do a `cargo clean && cargo build`, `git status` will show that the workspace is in dirty state.

File: build.rs (hunk 2 above)
Comment (nrc) [reply build.rs]: I was planning on landing this first, then #349 as a follow-up

**Comment by nrc** [comment]
> Another thought: it may be bad to modify the source directory

@BusyJay agree. I address this in #349 by making code generation opt-in, so a regular `cargo build` does not rebuild. I thought this is better than generating into the OUT_DIR because it means you can see the code and that might be helpful when debugging and so forth.

**Comment by ice1000** [comment]
`prost` support will generate codes in OUT_DIR and use `include!` macro to get them compiled.

File: build.rs (hunk 2 above)
Comment (ice1000) [reply build.rs]: Cool!

**Comment by brson** [comment]
Awesome start @nrc.

**Comment by BusyJay** [comment]
I'm not familiar with how cargo compile dependencies. If I compile several projects in parallel, which all depend on kvproto, will kvproto also be compiled in parallel? So will the source files get written concurrently? Is it a race condition that can lead to corrupted files?

**Comment by BusyJay** [comment]
> ...when debugging and so forth

Do you mean to debug it in the dependent project or inside kvproto? If debugging in dependent project, it may be easier to look at when the code is under target directory instead of $HOME/.cargo.

**Comment by nrc** [comment]
> I'm not familiar with how cargo compile dependencies. If I compile several projects in parallel, which all depend on kvproto, will kvproto also be compiled in parallel? So will the source files get written concurrently? Is it a race condition that can lead to corrupted files?

It will be rebuilt for each project in the project's target directory - compiled code is not shared between projects, although there is work in the pipeline to do so

> Do you mean to debug it in the dependent project or inside kvproto? If debugging in dependent project, it may be easier to look at when the code is under target directory instead of $HOME/.cargo.

Both. And actually just writing code as a downstream user - it is nice to look on GitHub and see the generated code.

**Comment by BusyJay** [comment]
> compiled code is not shared between projects, 

What I mean is the generated code, as they are flushed into the source directory.

**Comment by nrc** [comment]
> What I mean is the generated code, as they are flushed into the source directory.

Oh I see, if there are several projects using kvptoto via a `path` dependency, then I think you could have issues. But I don't think that should happen (it would mean a `path` dep which points outside of its own repo), so I don't think we should worry.

In any case #349 changes things to opt-in regeneration so it will not be an issue. If you like we can close this PR and discuss #349, but I thought it would be easier to land one piece at a time.

**Comment by nrc** [comment]
@BusyJay comment addressed and rebased PTAL

**Pull request merged**